### PR TITLE
fix(core): rank units by bias-corrected utility in Continual Backprop

### LINKED ISSUE #2343: CBP replaces units by the raw utility EMA, not the paper's bias-corrected utility (Dohare et al. 2024, Eq. 8) — it re-replaces its own freshly reset units
## Summary

`alberta_framework/core/continual_backprop.py::_select_replacement_index` ranks mature units by the **raw** utility EMA. The algorithm it cites ranks by the **bias-corrected** EMA.

Dohare, Hernandez-Garcia, Lan, Rahman, Mahmood & Sutton (2024), *"Loss of plasticity in deep continual learning"*, **Nature 632, 768–774**, Eq. 7–8:

```
u[l,i,t]     = eta * u[l,i,t-1] + (1 - eta) * y[l,i,t]        (7)
u_hat[l,i,t] = u[l,i,t] / (1 - eta ** a[l,i,t])               (8)
```

`a[l,i,t]` is the **per-unit age**. Algorithm 1 then selects the eligible units "with the smallest utility", i.e. smallest `u_hat`.

The authors' released implementation agrees — https://github.com/shibhansh/loss-of-plasticity, `lop/algos/gnt.py`:

```python
# GnT.update_utility
bias_correction = 1 - self.decay_rate ** self.ages[layer_idx]
...
self.bias_corrected_util[layer_idx] = self.util[layer_idx] / bias_correction

# GnT.test_features
new_features_to_replace = torch.topk(-self.bias_corrected_util[i][eligible_feature_indices],
                                     num_new_features_to_replace)[1]
```

`lop/algos/cbp_linear.py` and `lop/algos/cbp_conv.py` use the same per-unit-age correction.

## Why it does not cancel

The correction is per unit, and `_replace_one_unit` sets a replaced unit's `age` **and** `utility` back to 0. So eligible units in the same layer genuinely differ in age and therefore in warm-up bias. At the module defaults (`decay_rate=0.99`, `maturity_threshold=100`) a unit that has just become eligible reads at `1 - 0.99**100 = 0.6340` of its true utility, while a long-lived unit reads at ~1.0. Ranking on the raw EMA therefore systematically targets the youngest eligible units — precisely the units the maturity threshold exists to protect. The paper states the motivation explicitly:

> "initializing the outgoing weight to zero makes the new unit vulnerable to immediate reinitialized as it has zero utility. To protect new units from a reinitialization for maturity threshold m number of updates."

The maturity threshold only defers the problem by `m` steps; without Eq. 8 the freshly reset unit is still the lowest-ranked unit the moment it matures.

## Falsifier (deterministic, pristine `main` @ `c9aba7b5`)

Both scenarios use `decay_rate=0.99`, `maturity_threshold=100`, and drive the real `update_utility` EMA.

**A — young-strong vs old-weak.** Unit 0 has seen a constant contribution of `1.0` for 100 steps; unit 1 a constant `0.9` for 700 steps.

```
raw EMA        : unit0=0.633967 (age 100)   unit1=0.899209 (age 700)
Eq. 8 corrected: unit0=1.000000             unit1=0.900001
true steady-state contributions: unit0=1.0  unit1=0.9

main            REPLACED UNIT: [0]   <- the strictly more useful unit
Eq. 8 / gnt.py  REPLACED UNIT: [1]
```

**B — repeat victim.** Four units with *identical* true contribution `1.0`; unit 2 was reset 100 steps ago (i.e. it is the unit CBP replaced last time and has only just matured).

```
raw EMA        : [0.999997, 0.999997, 0.633967, 0.999997]
Eq. 8 corrected: [0.999997, 0.999997, 1.000000, 0.999997]

main            REPLACED UNIT: [2]   <- re-replaces its own fresh unit
Eq. 8 / gnt.py  REPLACED UNIT: [0]
```

## Scope

- `core/continual_backprop.py` is **not** a registered evidence source (`alberta_framework/evaluation/*` registers `average_reward`, `baseline_optimizers`, `canonical_upgd`, `ftl_world_model`, `intelligence_amplification`, `interaction_features`, `normalizers`, `oak`, `options`, `optimizers`, `types`, `update_safety` — not this file). `alberta-evidence-status` reports byte-identical per-claim validity before and after.
- `benchmarks/ipmnist_screening.py` carries its own CBP implementation and does not import `core.continual_backprop`, so no screening shard, summary, or incumbent number moves.
- In-tree consumers are the public `alberta_framework` API (the robot track's `core/{...,continual_backprop,...}` import) and the tests.

## Precedent

Same defect class as #480 (`fix(features): bias-correct EMA utility outputs`, merged) and issue #431, which fixed the warm-up debias for `future_utility` / `feature_discovery`. `core/continual_backprop.py` was not covered by that change.

## Not changed here (separate, smaller)

Eligibility uses `age >= maturity_threshold`; the paper says "units with age **more than** m" and `gnt.py` uses `torch.where(self.ages > self.maturity_threshold)`. That is a one-step off-by-one and a separate variable; deliberately left out of the fix so this change has exactly one moving part.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2343

`_select_replacement_index` in `alberta_framework/core/continual_backprop.py` previously ranked mature units using the raw EMA utility instead of the per-unit bias-corrected utility $\hat{u} = u / (1 - \eta^a)$ specified by Dohare et al. (2024, Eq. 8). Because freshly mature units have small ages ($a \approx m$), their raw EMA utility is heavily attenuated by warm-up bias ($1 - 0.99^{100} \approx 0.634$), causing CBP to systematically target and replace the youngest eligible units repeatedly.

### Changes
1. Updated `_select_replacement_index` to compute per-unit bias correction $1 - \text{decay}^{\text{age}}$ and rank eligible mature units by bias-corrected utility $u / (1 - \text{decay}^{\text{age}})$.
2. Forwarded `decay_rate=config.decay_rate` from `replace_units_with_flags` into `_select_replacement_index`.
3. Added unit test `test_select_replacement_index_uses_bias_corrected_utility` in `tests/test_continual_backprop.py` verifying that young mature units with higher true contributions are protected from replacement over older units with lower true utility.

### Verification
- `pytest tests/test_continual_backprop.py`: 96/96 tests passed.
- `ruff check`: All checks passed.
- `mypy`: Clean.

### CODE REVIEW

**Review by saltwire-dev — CHANGES_REQUESTED** [review]
## Request changes at exact head `68aa63d02e86fd984730afd858180c6f69871fe9`

Reviewed at exact head `68aa63d02e86fd984730afd858180c6f69871fe9` against `alberta_framework/core/continual_backprop.py` and sibling `bias_correct_future_utility` in `future_utility.py` at that SHA. Conversation sidebar is No reviews; the github-actions “produced no report” note is a comment, not a review.

Ranking mature units by `u / (1 - η^a)` is the right repair for #2343 — on the issue’s numbers a 100-step unit with true contribution 1.0 really does read at `1-0.99**100 ≈ 0.63397` while a 700-step unit with true 0.9 reads at `≈ 0.89921`, so raw `argmin` replaces the more useful unit. The hole is that this PR reimplements Eq. 8 inline instead of using the sibling helper #2343 named as the precedent, and the new test never drives the public replacement path the issue measured.

**Blocking: not the sibling `bias_correct_future_utility` contract #2343 pointed at**

Issue #2343 is explicit that this is the same defect class as merged #480 / #431, which already debiases ranking EMAs in `feature_discovery.py` via `bias_correct_future_utility`. At this SHA that helper is:

```python
def bias_correct_future_utility(utilities, ages, utility_decay, mode: str):
    age_count = jnp.maximum(ages.astype(jnp.float32), 0.0)
    decay = jnp.asarray(utility_decay, dtype=utilities.dtype)
    debias = 1.0 - jnp.power(decay, age_count)
    safe_debias = jnp.maximum(debias, jnp.asarray(1e-30, dtype=utilities.dtype))
    corrected = utilities / safe_debias
    return jnp.where(age_count > 0.0, corrected, utilities)
```

Floor is **1e-30**. Age-zero slots retain raw value. Ranking callers in `feature_discovery.py` go through this helper.

This PR inlines a different contract in `_select_replacement_index`:

```python
decay = jnp.asarray(decay_rate, dtype=jnp.float32)
bias_correction = 1.0 - jnp.power(decay, age.astype(jnp.float32))
safe_bias_correction = jnp.maximum(bias_correction, jnp.asarray(1e-12, dtype=jnp.float32))
bias_corrected_utility = utility / safe_bias_correction
mature = age >= jnp.asarray(maturity_threshold, dtype=age.dtype)
eligible = mature & jnp.isfinite(bias_corrected_utility)
```

Three divergences:

1. **Floor is 1e-12, not 1e-30.** This PR divides *every* unit, including age 0, by 1e-12. A leftover `u=1e-6` at age 0 becomes `û=1e6` here and stays `1e-6` in the sibling. Config allows `maturity_threshold >= 0`; this file already tests `maturity=0`.
2. **No age-0 identity.** Sibling: `where(age_count > 0, corrected, utilities)`. After `_replace_one_unit` this module sets `utilities=0` and `ages=0`. The next `replace_units_with_flags` without an intervening `update_utility` then sees age-0 zeros as `0/1e-12=0` and immediately re-picks them.
3. **Eligibility moved off the raw EMA.** Documented fail-closed path is `mature & isfinite(utility)`. Ranking on `isfinite(bias_corrected_utility)` after a 1e-12 floor means a raw 0 at age 0 is eligible, and a raw finite that overflows `u/1e-12` to inf is *not*. #2343 asked that this change have exactly one moving part.

`decay_rate: float = 0.99` is a fourth footgun. The sibling takes decay as a required argument. Production `replace_units_with_flags` does pass `config.decay_rate`, but the existing in-file caller at `tests/test_continual_backprop.py:222-225` does not (`_select_replacement_index(..., 10)` — default η=0.99, EMA was 0.9).

**Required fix:** import the sibling (mode `"age"`) or copy its exact contract: floor `1e-30`, `where(age > 0, u/(1-η^a), u)`, eligibility still `mature & isfinite(utility)`. Make `decay_rate` required.

```python
from alberta_framework.core.future_utility import bias_correct_future_utility

corrected = bias_correct_future_utility(utility, age, decay_rate, "age")
mature = age >= jnp.asarray(maturity_threshold, dtype=age.dtype)
eligible = mature & jnp.isfinite(utility) & jnp.isfinite(corrected)
masked = jnp.where(eligible, corrected, jnp.inf)
```

**Blocking: tests pin the closed form to itself, not the public replacement path**

Issue #2343’s reproduction is drive real `update_utility` EMA then run replacement (young-strong vs old-weak; repeat victim). The new test never calls `update_utility` or `replace_units_with_flags` / `maybe_replace_units` / `CBPMultiHeadMLPLearner.update`. It constructs `float(1.0 - 0.99**100)` vs `float(0.9 * (1.0 - 0.99**700))` and asserts `_select_replacement_index` flips them. That is the numbers the helper was written to invert, not the public CBP loop. Scenario B is untested. Equal-age tests still pass under *either* ranking.

**Required fix:** parametrize `replace_units_with_flags` with utilities/ages from `update_utility`, not `float(1-0.99**n)`. Pin scenario A (replace unit 1, leave unit 0) and scenario B (do not re-replace the just-matured unit). Pin age 0 with `m=0` against the sibling identity.

**Verified**

- Human PR; Conversation sidebar is literally No reviews.
- Head SHA `68aa63d02e86fd984730afd858180c6f69871fe9` matches the `.patch` From-line.
- `replace_units_with_flags` does thread `decay_rate=config.decay_rate`.
- Default-config arithmetic (`η=0.99`, `m=100`) does not hit the 1e-12 floor (`1-0.99**100 ≈ 0.634`). The hole is the sibling contract and the untested public path.

I would approve a follow-up that uses `bias_correct_future_utility` and drives `replace_units_with_flags` with real EMAs. Until then this is a third spelling of Eq. 8 that the public replacement path never exercises.

**Comment by lalalune** [comment]
Closing as superseded by merged #2345. The merged version applies per-unit EMA bias correction through the real replacement path and checks corrected utility eligibility. The replacement regression fails on main and the CBP/options panel passes all 206 tests.
